pkg/controllers/management/kontainerdrivermetadata/data_getter.go

```diff
@@ -0,0 +1,192 @@
+package kontainerdrivermetadata
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/rancher/rancher/pkg/settings"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/rancher/rancher/pkg/namespace"
+	"github.com/rancher/rke/util"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func GetRKESystemImages(k8sVersion string, sysImageLister v3.RKEK8sSystemImageLister, sysImages v3.RKEK8sSystemImageInterface) (v3.RKESystemImages, error) {
+	name := getName(k8sVersion)
+	sysImage, err := sysImageLister.Get(namespace.GlobalNamespace, name)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return v3.RKESystemImages{}, err
+		}
+		sysImage, err = sysImages.GetNamespaced(namespace.GlobalNamespace, name, metav1.GetOptions{})
+		if err != nil {
+			return v3.RKESystemImages{}, err
+		}
+	}
+	return sysImage.SystemImages, err
+}
+
+func GetRKEWindowsSystemImages(k8sVersion string, sysImageLister v3.RKEK8sWindowsSystemImageLister, sysImages v3.RKEK8sWindowsSystemImageInterface) (v3.WindowsSystemImages, error) {
+	name := getWindowsName(k8sVersion)
+	sysImage, err := sysImageLister.Get(namespace.GlobalNamespace, name)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return v3.WindowsSystemImages{}, err
+		}
+		sysImage, err = sysImages.GetNamespaced(namespace.GlobalNamespace, name, metav1.GetOptions{})
+		if err != nil {
+			return v3.WindowsSystemImages{}, err
+		}
+	}
+	return sysImage.SystemImages, err
+}
+
+func GetRKEAddonTemplate(k8sVersion string, addonName string, addonLister v3.RKEAddonLister, addons v3.RKEAddonInterface) (string, error) {
+	names := []string{
+		fmt.Sprintf("%s-%s", strings.ToLower(addonName), getName(k8sVersion)),
+		fmt.Sprintf("%s-%s", strings.ToLower(addonName), getName(util.GetTagMajorVersion(k8sVersion))),
+		fmt.Sprintf("%s-%s", strings.ToLower(addonName), "default"),
+	}
```

Comment:
Why do we have 3 name formats? where do we add the addon crds?

Reply:
adding in `createOrUpdateAddonCRD`, https://github.com/rancher/rancher/pull/20785/files#diff-e282be4392d4256b319d411aa68fbd79R310 

this is the preference order, if minor version specific template is available, we should use it, followed by major version and then last default.